sway-ast/src/token.rs

```diff
@@ -67,45 +67,83 @@ impl PunctKind {
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Hash)]
-pub struct GenericGroup<T> {
-    pub delimiter: Delimiter,
+pub struct DelimitedGroup<T> {
+    pub delimiters: Delimiters,
     pub token_stream: T,
     pub span: Span,
 }
 
-pub type Group = GenericGroup<TokenStream>;
-pub type CommentedGroup = GenericGroup<CommentedTokenStream>;
+pub type Group = DelimitedGroup<TokenStream>;
+pub type CommentedGroup = DelimitedGroup<CommentedTokenStream>;
 
-impl<T> Spanned for GenericGroup<T> {
+impl<T> Spanned for DelimitedGroup<T> {
     fn span(&self) -> Span {
         self.span.clone()
     }
 }
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Delimiters {
+    pub opening: Option<OpeningDelimiter>,
+    pub closing: Option<ClosingDelimiter>,
+}
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Delimiter {
+    pub kind: DelimiterKind,
+    pub span: Span,
+}
+impl Spanned for Delimiter {
+    fn span(&self) -> Span {
+        self.span.clone()
+    }
+}
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum DelimiterKind {
+    Opening(OpeningDelimiter),
+    Closing(ClosingDelimiter),
+}
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub enum Delimiter {
+pub enum OpeningDelimiter {
     Parenthesis,
-    Brace,
-    Bracket,
+    CurlyBrace,
+    SquareBracket,
+    AngleBracket,
 }
-
-impl Delimiter {
-    pub fn as_open_char(self) -> char {
+impl OpeningDelimiter {
+    pub fn as_char(&self) -> char {
         match self {
-            Delimiter::Parenthesis => '(',
-            Delimiter::Brace => '{',
-            Delimiter::Bracket => '[',
+            OpeningDelimiter::Parenthesis => '(',
+            OpeningDelimiter::CurlyBrace => '{',
+            OpeningDelimiter::SquareBracket => '[',
+            OpeningDelimiter::AngleBracket => '<',
         }
     }
-    pub fn as_close_char(self) -> char {
+    pub fn get_closing_delimiter(&self) -> ClosingDelimiter {
         match self {
-            Delimiter::Parenthesis => ')',
-            Delimiter::Brace => '}',
-            Delimiter::Bracket => ']',
+            Self::Parenthesis => ClosingDelimiter::Parenthesis,
+            Self::CurlyBrace => ClosingDelimiter::CurlyBrace,
+            Self::SquareBracket => ClosingDelimiter::SquareBracket,
+            Self::AngleBracket => ClosingDelimiter::AngleBracket,
+        }
+    }
+}
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
```

Comment:
This enum has the same variants as `OpeningDelimiter`. I feel like we should be able to keep the original `Delimiter` enum and make that work. Perhaps we could rename it to. 

```rust
pub enum DelimiterType {
    Parenthesis,
    CurlyBrace,
    SquareBracket,
    AngleBracket,
}
```